pkg/controllers/jobset_controller.go

```diff
@@ -556,6 +569,51 @@ func (r *JobSetReconciler) reconcileReplicatedJobs(ctx context.Context, js *jobs
 	return nil
 }
 
+func (r *JobSetReconciler) recordJobCreationFailed(js *jobset.JobSet, err error) {
+	key := types.NamespacedName{Namespace: js.GetNamespace(), Name: js.GetName()}.String()
+	r.recordThrottledEvent(key, js, js.Spec.ReplicatedJobs, corev1.EventTypeWarning, constants.JobCreationFailedReason, err.Error())
+}
+
+type lastEventCacheRecord struct {
+	Timestamp time.Time
+	Opaque    any
+}
+
+func (r *JobSetReconciler) recordThrottledEvent(eventKey string, object runtime.Object, opaque any, eventType, reason, messageFmt string, args ...interface{}) {
+	r.eventTimestampLock.Lock()
+	v, ok := r.eventTimestamps.Get(eventKey)
+	r.eventTimestamps.Add(eventKey, lastEventCacheRecord{
+		Timestamp: r.clock.Now(),
+		Opaque:    opaque,
+	})
+	r.eventTimestampLock.Unlock()
+
+	if ok {
+		record := v.(lastEventCacheRecord)
+		if time.Since(record.Timestamp) < eventThrottlePeriod && reflect.DeepEqual(record.Opaque, opaque) {
+			return
+		}
+	}
+
+	r.Record.Eventf(object, eventType, reason, messageFmt, args...)
+}
+
+func (r *JobSetReconciler) RunCleaner(ctx context.Context) {
+	ticker := time.NewTicker(eventTimestampCacheResetPeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			r.eventTimestampLock.Lock()
+			r.eventTimestamps.Clear()
```

Comment:
Not possible to implement a more clever mechanism with the LRU cache, but just resetting it periodically works just fine, I think. In the worst case you get 2 events very close to each other, who cares.